tensorrt_llm/_torch/custom_ops/torch_custom_ops.py

```diff
@@ -1035,6 +1042,235 @@ def _(
+        inputs: List[torch.Tensor],
+        tactic,
+    ) -> torch.Tensor:
+        """Performs fp8 blockwise (deepgemm like) operation using CuTe DSL.
+        :param a (inputs[0]): Input tensor of shape (m, k)
+        :type a: torch.Tensor, type: fp4
+        :param b (inputs[1]): Weight tensor of shape (n, k)
+        :type b: torch.Tensor, type: fp4
+        :param a_sf (inputs[2]): Input scale tensor of shape (k//16, m).
+        :type a_sf: torch.Tensor, type: fp8
+        :param b_sf (inputs[3]): Weight scale tensor of shape (n, k//16)
+        :type b_sf: torch.Tensor, type: fp8
+        :return: Output tensor of shape (m, n)
+        :rtype: torch.Tensor, type: bf16
+        """
+        sf_vec_size = 16
+
+        if isinstance(tactic, tuple):
+            mma_tiler_mn, cluster_shape_mn = tactic
+        else:
+            # fallback to default tactic
+            mma_tiler_mn, cluster_shape_mn = [
+                (128, 128),
+                (1, 1),
+            ]
+
+        a_tensor, b_tensor, a_sf_tensor, b_sf_tensor, alpha, output_dtype = inputs
+        assert output_dtype == torch.bfloat16
+        m, k, n = a_tensor.shape[0], a_tensor.shape[1], b_tensor.shape[0]
+        c_tensor = torch.empty(*(m, n), dtype=output_dtype, device="cuda")
+
+        real_k = k * 2
+        sf_m = pad_up(m, 128)
+        sf_k = pad_up(real_k // sf_vec_size, 4)
+        sf_n = pad_up(n, 128)
+
+        a_ptr = make_ptr(
+            # cutlass type
+            cutlass.Float4E2M1FN,
+            a_tensor.data_ptr(),
+            cute.AddressSpace.gmem,
+            assumed_align=32,
+        )
+        b_ptr = make_ptr(
+            cutlass.Float4E2M1FN,
+            b_tensor.data_ptr(),
+            cute.AddressSpace.gmem,
+            assumed_align=32,
+        )
+        a_sf_ptr = make_ptr(
+            cutlass.Float8E4M3FN,
+            a_sf_tensor.data_ptr(),
+            cute.AddressSpace.gmem,
+            assumed_align=16,
+        )
+        b_sf_ptr = make_ptr(
+            cutlass.Float8E4M3FN,
+            b_sf_tensor.data_ptr(),
+            cute.AddressSpace.gmem,
+            assumed_align=16,
+        )
+        c_ptr = make_ptr(
+            cutlass.BFloat16,
+            c_tensor.data_ptr(),
+            cute.AddressSpace.gmem,
+            assumed_align=16,
+        )
+
+        # get stream
+        torch_stream = torch.cuda.current_stream()
+        stream = cuda.CUstream(torch_stream.cuda_stream)
+
+        gemm_wrapper_func = Sm100BlockScaledPersistentDenseGemmKernelWrapper
+        cache_key = (
+            sf_vec_size,
+            mma_tiler_mn,
+            cluster_shape_mn,
+        )
+        if cache_key not in CuteDSLNVFP4BlackwellLinear.kernel_dict:
+            gemm = gemm_wrapper_func(
+                sf_vec_size,
+                mma_tiler_mn,
+                cluster_shape_mn,
+            )
+            # Compute max active clusters on current device
+            hardware_info = cutlass.utils.HardwareInfo()
```

Comment:
We should cache these I think to reduce the host code. Will this call CUDA API everytime to query the HW info?